File: src/crates/ncode-python/python/ncode/_reader.py
"""Zero-copy reader for the NCB ("Ncode Columnar Buffer") wire format.

The native extension returns query results as NCB ``bytes``. This module parses
the fixed header + directory and exposes each column. Numeric columns are
surfaced via ``memoryview.cast`` *views* over the transferred buffer — no
per-value copy — which is possible because the encoder 8-byte-aligns every
numeric payload.
"""

from __future__ import annotations

import struct
from dataclasses import dataclass
from enum import IntEnum
from typing import Any, Optional, Sequence

_MAGIC = b"NCB1"
_HEADER_LEN = 24
_COLDIR_LEN = 40


class DataType(IntEnum):
    INT64 = 0
    FLOAT64 = 1
    BOOL = 2
    UTF8 = 3


def _valid_at(validity: Optional[memoryview], row: int) -> bool:
    if validity is None:
        return True
    return bool(validity[row >> 3] & (1 << (row & 7)))


@dataclass
class Column:
    """A single decoded column."""

    name: str
    type: DataType
    length: int
    _validity: Optional[memoryview]
    _values: Sequence[Any]  # a memoryview (numeric/bool) or a decoded helper

    def is_valid(self, row: int) -> bool:
        return _valid_at(self._validity, row)

    def get(self, row: int) -> Any:
        if row < 0 or row >= self.length or not self.is_valid(row):
            return None
        return self._values[row]

    def to_list(self) -> list[Any]:
        return [self.get(i) for i in range(self.length)]


class _BoolView:
    """Bit-unpacking view over a bool bitmap."""

    def __init__(self, bits: memoryview):
        self._bits = bits

    def __getitem__(self, row: int) -> bool:
        return bool(self._bits[row >> 3] & (1 << (row & 7)))


class _Utf8View:
    """Offset-indexed view over a UTF-8 char-data buffer."""

    def __init__(self, offsets: Sequence[int], data: memoryview):
        self._offsets = offsets
        self._data = data

    def __getitem__(self, row: int) -> str:
        start, end = self._offsets[row], self._offsets[row + 1]
        return bytes(self._data[start:end]).decode("utf-8")


class Batch:
    """A decoded columnar result set."""

    def __init__(self, num_rows: int, columns: list[Column]):
        self.num_rows = num_rows
        self.columns = columns
        self._by_name = {c.name: c for c in columns}

    def column(self, name: str) -> Optional[Column]:
        return self._by_name.get(name)

    def to_rows(self) -> list[dict[str, Any]]:
        return [
            {c.name: c.get(r) for c in self.columns} for r in range(self.num_rows)
        ]

    def __repr__(self) -> str:
        cols = ", ".join(f"{c.name}:{c.type.name.lower()}" for c in self.columns)
        return f"<ncode.Batch rows={self.num_rows} [{cols}]>"


def decode_batch(buf: bytes) -> Batch:
    """Decode an NCB buffer into a :class:`Batch`."""
    mv = memoryview(buf)
    if mv[0:4] != _MAGIC:
        raise ValueError("not an NCB buffer (bad magic)")
    version = struct.unpack_from("<H", mv, 4)[0]
    if version != 1:
        raise ValueError(f"unsupported NCB version {version}")

    num_columns, num_rows, dir_off = struct.unpack_from("<III", mv, 8)

    columns: list[Column] = []
    for c in range(num_columns):
        d = dir_off + c * _COLDIR_LEN
        name_off, name_len = struct.unpack_from("<II", mv, d)
        dtype = DataType(mv[d + 8])
        has_validity = bool(mv[d + 9] & 1)
        validity_off, validity_len = struct.unpack_from("<II", mv, d + 12)
        buf1_off, _buf1_len, buf2_off, buf2_len = struct.unpack_from("<IIII", mv, d + 20)

        name = bytes(mv[name_off : name_off + name_len]).decode("utf-8")
        validity = (
            mv[validity_off : validity_off + validity_len] if has_validity else None
        )

        values: Sequence[Any]
        if dtype is DataType.INT64:
            # Zero-copy: reinterpret the aligned little-endian region as int64.
            values = mv[buf1_off : buf1_off + num_rows * 8].cast("q")
        elif dtype is DataType.FLOAT64:
            values = mv[buf1_off : buf1_off + num_rows * 8].cast("d")
        elif dtype is DataType.BOOL:
            n_bytes = (num_rows + 7) // 8
            values = _BoolView(mv[buf1_off : buf1_off + n_bytes])
        elif dtype is DataType.UTF8:
            offsets = struct.unpack_from(f"<{num_rows + 1}I", mv, buf1_off)
            data = mv[buf2_off : buf2_off + buf2_len]
            values = _Utf8View(offsets, data)
        else:  # pragma: no cover - guarded by DataType()
            raise ValueError(f"unsupported NCB type code {dtype}")

        columns.append(Column(name, dtype, num_rows, validity, values))

    return Batch(num_rows, columns)

File: src/crates/ncode-node/ts/reader.ts
/**
 * Zero-copy reader for the NCB ("Ncode Columnar Buffer") wire format.
 *
 * The native addon returns query results as an NCB `Buffer`. This reader
 * parses the fixed header + directory and exposes each column. Numeric columns
 * are surfaced as typed-array *views* directly over the transferred bytes
 * (`Float64Array` / `BigInt64Array`) — no per-value copy — whenever the buffer
 * lands on an 8-byte boundary, which the encoder guarantees for external
 * buffers.
 */

const MAGIC = 0x3142434e; // "NCB1" (0x4E 0x43 0x42 0x31) read as little-endian u32
const HEADER_LEN = 24;
const COLDIR_LEN = 40;

export enum DataType {
  Int64 = 0,
  Float64 = 1,
  Bool = 2,
  Utf8 = 3,
}

export type Scalar = bigint | number | boolean | string | null;

export interface NcodeColumn {
  readonly name: string;
  readonly type: DataType;
  readonly length: number;
  isValid(row: number): boolean;
  get(row: number): Scalar;
  /** Materialize the whole column to a JS array (copies). */
  toArray(): Scalar[];
}

const decoder = new TextDecoder();

/** Read a validity bit (LSB-first). `undefined` bitmap => all valid. */
function validAt(bitmap: Uint8Array | undefined, row: number): boolean {
  if (!bitmap) return true;
  return (bitmap[row >> 3] & (1 << (row & 7))) !== 0;
}

class ColumnImpl implements NcodeColumn {
  constructor(
    readonly name: string,
    readonly type: DataType,
    readonly length: number,
    private readonly validity: Uint8Array | undefined,
    private readonly readRaw: (row: number) => Scalar,
  ) {}

  isValid(row: number): boolean {
    return validAt(this.validity, row);
  }

  get(row: number): Scalar {
    if (row < 0 || row >= this.length || !this.isValid(row)) return null;
    return this.readRaw(row);
  }

  toArray(): Scalar[] {
    const out = new Array<Scalar>(this.length);
    for (let i = 0; i < this.length; i++) out[i] = this.get(i);
    return out;
  }
}

export interface NcodeBatch {
  readonly numRows: number;
  readonly columns: NcodeColumn[];
  column(name: string): NcodeColumn | undefined;
  /** Row-oriented view (copies) — convenient for small result sets. */
  toRows(): Record<string, Scalar>[];
}

/** Decode an NCB buffer into a columnar batch. */
export function decodeBatch(buf: Uint8Array): NcodeBatch {
  const ab = buf.buffer;
  const base = buf.byteOffset;
  const view = new DataView(ab, base, buf.byteLength);

  if (view.getUint32(0, true) !== MAGIC) {
    throw new Error("not an NCB buffer (bad magic)");
  }
  const version = view.getUint16(4, true);
  if (version !== 1) throw new Error(`unsupported NCB version ${version}`);

  const numColumns = view.getUint32(8, true);
  const numRows = view.getUint32(12, true);
  const dirOff = view.getUint32(16, true);

  // Can we make an aligned typed-array view at absolute offset `off`?
  const aligned = (off: number) => ((base + off) & 7) === 0;

  const columns: NcodeColumn[] = [];
  for (let c = 0; c < numColumns; c++) {
    const d = dirOff + c * COLDIR_LEN;
    const nameOff = view.getUint32(d, true);
    const nameLen = view.getUint32(d + 4, true);
    const dtype = view.getUint8(d + 8) as DataType;
    const hasValidity = (view.getUint8(d + 9) & 1) !== 0;
    const validityOff = view.getUint32(d + 12, true);
    const validityLen = view.getUint32(d + 16, true);
    const buf1Off = view.getUint32(d + 20, true);
    const buf2Off = view.getUint32(d + 28, true);
    const buf2Len = view.getUint32(d + 32, true);

    const name = decoder.decode(new Uint8Array(ab, base + nameOff, nameLen));
    const validity = hasValidity
      ? new Uint8Array(ab, base + validityOff, validityLen)
      : undefined;

    let readRaw: (row: number) => Scalar;
    switch (dtype) {
      case DataType.Int64: {
        if (aligned(buf1Off)) {
          const arr = new BigInt64Array(ab, base + buf1Off, numRows);
          readRaw = (r) => arr[r];
        } else {
          const dv = view;
          readRaw = (r) => dv.getBigInt64(buf1Off + r * 8, true);
        }
        break;
      }
      case DataType.Float64: {
        if (aligned(buf1Off)) {
          const arr = new Float64Array(ab, base + buf1Off, numRows);
          readRaw = (r) => arr[r];
        } else {
          const dv = view;
          readRaw = (r) => dv.getFloat64(buf1Off + r * 8, true);
        }
        break;
      }
      case DataType.Bool: {
        const bits = new Uint8Array(ab, base + buf1Off, Math.ceil(numRows / 8));
        readRaw = (r) => (bits[r >> 3] & (1 << (r & 7))) !== 0;
        break;
      }
      case DataType.Utf8: {
        const offsets = new Uint32Array(numRows + 1);
        for (let i = 0; i <= numRows; i++) {
          offsets[i] = view.getUint32(buf1Off + i * 4, true);
        }
        const data = new Uint8Array(ab, base + buf2Off, buf2Len);
        readRaw = (r) => decoder.decode(data.subarray(offsets[r], offsets[r + 1]));
        break;
      }
      default:
        throw new Error(`unsupported NCB type code ${dtype}`);
    }

    columns.push(new ColumnImpl(name, dtype, numRows, validity, readRaw));
  }

  return {
    numRows,
    columns,
    column(name) {
      return columns.find((col) => col.name === name);
    },
    toRows() {
      const rows: Record<string, Scalar>[] = [];
      for (let r = 0; r < numRows; r++) {
        const row: Record<string, Scalar> = {};
        for (const col of columns) row[col.name] = col.get(r);
        rows.push(row);
      }
      return rows;
    },
  };
}
